lib/queryBuilder/QueryBuilder.js

```diff
@@ -1070,11 +1070,11 @@ function beforeExecute(builder) {
   // Resolve all before hooks before building and executing the query
   // and the rest of the hooks.
   promise = chainOperationHooks(promise, builder, 'onBefore1');
+  promise = chainOperationHooks(promise, builder, 'onBefore2');
 
   promise = chainHooks(promise, builder, builder.context().runBefore);
   promise = chainHooks(promise, builder, builder.internalContext().runBefore);
 
-  promise = chainOperationHooks(promise, builder, 'onBefore2');
```

Comment:
I am a bit afraid of this change - could you tell more about why the execution order of hooks had to be changed to fix the issue, please?

Reply:
I need `runBefore` to run after `onBefore2()` which triggers the execution of `$beforeUpdate()`. I tried it an it didn't break any tests. I looked at `afterExecute()`, and it has the same sequence.

In the end, this is just a proof of concept more than anything for now. I do thin that the behavior it produces is more according to what I would expect though. 

```js
function afterExecute(builder, result) {
  let promise = Promise.resolve(result);

  promise = chainOperationHooks(promise, builder, 'onAfter1');
  promise = chainOperationHooks(promise, builder, 'onAfter2');

  promise = chainHooks(promise, builder, builder.context().runAfter);
  promise = chainHooks(promise, builder, builder.internalContext().runAfter);

  promise = chainOperationHooks(promise, builder, 'onAfter3');

  return promise;
}
```